### ISSUE TITLE
[Suggestion] Shotgun Push Fix as a standalone mod

### ISSUE THREAD

**Issue by Lordmau5** [Crackdown-PD2/deathvox#114]
Hey there!

So as everyone knows PAYDAY 2 is quite the bugged mess in certain spots, and this is definitely one of them: *Shotgun pushes / launches not syncing with other clients*

I've asked in the ModWorkshop Discord if someone might know a fix for this or if that's just something that we'll never see fixed and they linked me to this repository and 2 commits.

Now, I'm sure some people would like to play the game without all the other cool things Crackdown brings to the table and mostly the shotgun launch bug being fixed so I'm curious:

Would there be a chance that you guys could release a standalone version of just that as a mod?
Alternatively, if I could get permission (and would obviously credit you guys for that), I could release it as a mod on the ModWorkshop website, too.

If this isn't the correct place to ask about that then I'm sorry - I just couldn't find another possibility to reach out to you guys.

**Comment by hoxiwoopboop** [comment]
It depends. We didn't exactly test if the receiver also needs to have the same type of delivery (in this case, "ordered" instead of vanilla's "reliable") for this to work with no issues.

I'm assuming it can be used locally to sync the push, as the order/queue for syncing is determined by the sender from what I understand when reading the relevant network handler code, but until I'm certain, I'd refrain from making it a standalone mod.

To give you an idea, using send_to_peers_synched when the delivery isn't "ordered" results in breaking most of the network syncing for other clients. As your queue gets stuck and anything that uses "ordered" won't be sent at ALL, making those other clients get stuck in a limbo, to put it simple. This could also happen if the receiver doesn't have the delivery to be the same, but we haven't tested yet.

Lastly, since I was able (with some help) to pin-point the cause of the issue, I talked to one of the devs and they're fully aware of what causes it and how to fix it. While I don't know what they'll do, it's likely that this will be fixed in the next update of the base game.

**Comment by Lordmau5** [comment]
I've done some testing with a friend before that didn't have it installed and there were no issues.

Well, there was one issue - when *he* was launching enemies it wasn't syncing over to me, but when *I* was doing it he had them get launched properly.

And yeah, I've been informed about one of you guys reaching out to OVK about this (I brought it up as well in a mail that I sent to them today, together with some more stuff).

Also, is there a reason why `send_to_peers_synched` is used over `send_to_peers` (which, from my understanding, is an async one)?

**Comment by hoxiwoopboop** [comment]
"I've done some testing with a friend before that didn't have it installed and there were no issues.

Well, there was one issue - when he was launching enemies it wasn't syncing over to me, but when I was doing it he had them get launched properly."

Yeah, this is what I assumed would happen. Only being relevant for the sender. This only affects how you send it (but I wasn't sure if the receiver also needed to have the correct delivery, as normally these aren't changed).

"Also, is there a reason why send_to_peers_synched is used over send_to_peers (which, from my understanding, is an async one)?"

It queues the sync so that it arrives in the correct order (in this case, enemy taking damage and dying first, shotgun push later), as this is what caused the thing to not work in the first place. It was always being synced, but it rarely was done in the correct order, so the push would mostly arrive before hand and do nothing as the enemy is still alive.

The ordered change did nothing on it's own. I've been checking with the network settings file and all functions that use "ordered" as delivery use send_to_peers_synched for this reason.

**Comment by Lordmau5** [comment]
> Yeah, this is what I assumed would happen. Only being relevant for the sender. This only affects how you send it (but I wasn't sure if the receiver also needed to have the correct delivery, as normally these aren't changed).

Well from my understanding the attribute for that sounds like it is only relevant for outgoing messages, not incoming ones, seeing as incoming ones, maybe?
On the other hand, the network event / message doesn't change - only the order / type of it - which means it **shouldn't** affect receivers at all. (Can probably still happen, but I can give more info after extensive testing this weekend)

Back to one of the other questions though:
Do you think that releasing it as a standalone mod would be an option / me releasing it with crediting you **AND** labeling it as experimental so that players (hopefully) won't be complaining about it? (I mean, we both know players will *still* complain, but that at least filters out those that can't read)

**Comment by hoxiwoopboop** [comment]
> Well from my understanding the attribute for that sounds like it is only relevant for outgoing messages, not incoming ones, maybe?

That's what I was thinking, yes.
> On the other hand, the network event / message doesn't change - only the order / type of it - which means it shouldn't affect receivers at all. (Can probably still happen, but I can give more info after extensive testing this weekend)

Again, I'm sure about the order, not so much about the type affecting the receiver if they have it different. Shouldn't be the case but I haven't tested.
> Back to one of the other questions though:
Do you think that releasing it as a standalone mod would be an option / me releasing it with crediting you AND labeling it as experimental so that players (hopefully) won't be complaining about it? (I mean, we both know players will still complain, but that at least filters out those that can't read)

I can look into making it a standalone mod but not right now, maybe tomorrow after I'm 100% sure it won't cause issues. I'm also not doing anything to make sure CopDamage:sync_damage_bullet won't affect other mods (if any mods modify that function at all), I can't really bother to spend time doing so.

**Comment by Lordmau5** [comment]
> I can look into making it a standalone mod but not right now, maybe tomorrow after I'm 100% sure it won't cause issues.

That sounds neat! Thanks in advance :)
In the meantime I'll be using a small mod that I made that only replaces those 2 `GamePlayCentralManager` methods and the network settings with a few friends (We wanted to do some speedruns and this would be beneficial for heists we are going to be doing in stealth for faster times)

> I'm also not doing anything to make sure CopDamage:sync_damage_bullet won't affect other mods (if any mods modify that function at all), I can't really bother to spend time doing so.

What exactly is that part doing? Is `CopDamage` being handled in a different manner over the actual shotgun pushing functionality / methods?
I mean, let's be real - shotgun pushing doesn't even matter when the heist is loud, as there's no need to have dead cops in specific positions, so in case that method is handling that, then that can probably be ignored.

**Comment by hoxiwoopboop** [comment]
[https://github.com/Crackdown-PD2/deathvox/commit/fb4d813349822bdce9cbf410787659331d751eab](url)

It makes sure synced bullet damage won't trigger a push during stealth (and lets the syncing feature handle it), which can now happen in vanilla due to some changes done in vanilla in the latest updates, for ``GameplayCentralManager`` and ``CopDamage`` (both bad and "good" changes). If you don't change that and add the sync fix, you can get super pushes during stealth.

[shotgunpushfixes reference (do not just extract in mods folder).zip](https://github.com/Crackdown-PD2/deathvox/files/3894767/shotgunpushfixes.reference.do.not.just.extract.in.mods.folder.zip)

As the file says, this is to be used for reference when editing the respective files.

**Comment by Lordmau5** [comment]
Been a while since I brought this up - I was checking the recent commits to the experimental branch today and noticed this:
https://github.com/Crackdown-PD2/deathvox/commit/2823463b652229575ee94dc2b9d84881f62f43d6

Does that mean there's no real need for the shotgun push fix anymore in that case? 🤔

**Comment by hoxiwoopboop** [comment]
You only need to make sure that send_to_peers_synched is used in gameplaycentralmanager instead of just send_to_peers. They pretty much got everything right but forgot to change that, so a push can still sync before the enemy actually gets killed.

Just grab the vanilla function and add that change pretty much.

I'll be updating more stuff here, including how the cosmetic copdamage push used outside of stealth gets the direction to make it more accurate. So be sure to check for this in a bunch of days.

**Comment by Lordmau5** [comment]
Ah alright, gotcha!

Also, did you guys ever think of releasing a standalone mod for just the shotgun push over the past month when I first brought it up?
Or would you, as I said, be fine with me releasing one?

**Comment by Lordmau5** [comment]
It's been a fair while and from some testing I did with a friend yesterday it seems like Overkill have managed to fix this long-standing bug in the base game

However... it still wasn't perfect when we shot the guards. Probably just due to some desync, but I'm not sure if that's something that could be fixed.

When I first tried it with a small mod I made with your code (which overwrote some of the functions to better handle the pushing) and I was too close to an enemy (or civilian) their model spazzed out and got teleported away.

Just thought I'd give a small update on this - I'm not sure if you guys ever removed that shotgun push code from your mod so there's that :D
